src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,395 @@
+#include "AddAtomicMutex.h"
+#include "ExprUsesVar.h"
+#include "Func.h"
+#include "IREquality.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "OutputImageParam.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::set;
+using std::string;
+
+namespace {
+
+/** Collect names of all stores inside a statement. */
+class CollectStoreNames : public IRGraphVisitor {
+public:
+    Scope<void> store_names;
+
+protected:
+    using IRGraphVisitor::visit;
+
+    void visit(const Store *op) override {
+        include(op->predicate);
+        include(op->value);
+        include(op->index);
+        store_names.push(op->name);
+    }
+};
+
+/** Find Store inside of an Atomic node and return their indices. */
+class FindStoreIndex : public IRGraphVisitor {
+public:
+    Expr index;
+
+protected:
+    using IRGraphVisitor::visit;
+
+    // Need to also extract the let bindings of a Store index.
+    void visit(const Let *op) override {
+        IRGraphVisitor::visit(op);
+        if (index.defined()) {
+            if (expr_uses_var(index, op->name)) {
+                index = Let::make(op->name, op->value, index);
+            }
+        }
+    }
+    void visit(const LetStmt *op) override {
+        IRGraphVisitor::visit(op);
+        if (index.defined()) {
+            if (expr_uses_var(index, op->name)) {
+                index = Let::make(op->name, op->value, index);
+            }
+        }
+    }
+
+    void visit(const Store *op) override {
+        // Ideally we want to insert equal() checks here for different stores,
+        // but the indices of them actually are different in the case of tuples,
+        // since they usually refer to the strides/min/extents of their own tuple
+        // buffers. However, different elements in a tuple would have the same
+        // strides/min/extents so we are fine.
+        if (index.defined()) {
+            return;
+        }
+        index = op->index;
+        op->value.accept(this);
+    }
+};
+
+/** Throw an assertion for cases where the indexing on left-hand-side of
+ *  an atomic update references to itself.
+ *  e.g. f(clamp(f(r), 0, 100)) = f(r) + 1 should be rejected. */
+class CheckAtomicValidity : public IRGraphVisitor {
+protected:
+    using IRVisitor::visit;
+
+    void visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+
+        // Find the indices from the Store nodes inside the body.
+        FindStoreIndex find;
+        op->body.accept(&find);
+
+        Expr index = find.index;
```

Comment:
If evaluating the index (including the containing lets) has side-effects, this will trigger them twice. Maybe lift the index computation into a letstmt outside of the atomic node.

Reply:
Fixed.